src/puya/awst_build/eb/compile.py

```diff
@@ -0,0 +1,262 @@
+                template_variables=template_vars,
+                wtype=self.produces.wtype,
+                source_location=location,
+            )
+        )
+
+
+class CompileLogicSigFunctionBuilder(FunctionBuilder):
+    def __init__(self, location: SourceLocation):
+        super().__init__(location)
+        self.produces = pytypes.CompiledLogicSigType
+
+    @typing.override
+    def call(
+        self,
+        args: Sequence[NodeBuilder],
+        arg_kinds: list[mypy.nodes.ArgKind],
+        arg_names: list[str | None],
+        location: SourceLocation,
+    ) -> InstanceBuilder:
+        type_reference, kwargs = _single_positional_and_kwargs(arg_names, args, location)
+        logic_sig = LogicSigReference("", "")
+        match type_reference:
+            case LogicSigExpressionBuilder(ref=logic_sig):
+                pass
+            case None:
+                pass
+            case _:
+                logger.error("invalid logicsig reference", location=type_reference.source_location)
+
+        prefix, template_vars = _extract_prefix_template_args(kwargs)
+        _expect_empty(kwargs.values())
+
+        return CompiledLogicSigExpressionBuilder(
+            CompiledLogicSig(
+                logic_sig=logic_sig,
+                prefix=prefix,
+                template_variables=template_vars,
+                wtype=self.produces.wtype,
+                source_location=location,
+            )
+        )
+
+
+def _extract_allocation_overrides(
+    name_args: dict[str, NodeBuilder]
+) -> Mapping[CompiledReferenceField, Expression]:
+    allocation_overrides = dict[CompiledReferenceField, Expression]()
+    for field in _ALLOCATION_OVERRIDE_FIELDS:
+        if arg := name_args.pop(field, None):
+            allocation_overrides[field] = expect.exactly_one_arg_of_type_else_dummy(
+                [arg], pytypes.UInt64Type, location=arg.source_location, resolve_literal=True
+            ).resolve()
+    return allocation_overrides
+
+
+def _extract_prefix_template_args(
+    name_args: dict[str, NodeBuilder]
+) -> tuple[str | None, Mapping[str, Expression]]:
+    prefix: str | None = None
+    template_vars: Mapping[str, Expression] = {}
+
+    if template_vars_node := name_args.pop("template_vars", None):
+        if isinstance(template_vars_node, DictLiteralBuilder):
+            template_vars = {k: v.resolve() for k, v in template_vars_node.mapping.items()}
+        else:
+            logger.error("unexpected argument type", location=template_vars_node.source_location)
+    if prefix_node := name_args.pop("template_vars_prefix", None):
+        prefix = expect.simple_string_literal(
+            prefix_node, default=expect.default_fixed_value(None)
+        )
+    return prefix, template_vars
+
+
+def _single_positional_and_kwargs(
+    names: Sequence[str | None], args: Sequence[NodeBuilder], location: SourceLocation
+) -> tuple[NodeBuilder | None, dict[str, NodeBuilder]]:
+    positional: NodeBuilder | None = None
```

Comment:
could use `utils.get_arg_mapping` instead of this function, and `_expect_empty` below would not be required then either